// cw-sdk-node/build/websocket/proto/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.periodNames = exports.orderTypeToProto = exports.orderTypeFromProto = exports.fundingTypeToProto = exports.fundingTypeFromProto = exports.privateOrderSideToProto = exports.privateOrderSideFromProto = void 0;
// NOTE: the mapping for buy/sell is different for PrivateOrder and PublicTrade
exports.privateOrderSideFromProto = {
    0: 'sell',
    1: 'buy'
};
exports.privateOrderSideToProto = {
    sell: 0,
    buy: 1
};
exports.fundingTypeFromProto = {
    0: 'spot',
    1: 'margin'
};
exports.fundingTypeToProto = {
    spot: 0,
    margin: 1
};
exports.orderTypeFromProto = {
    0: 'market',
    1: 'limit'
    // 2: "stoploss",
    // 3: "stoplosslimit",
    // 4: "takeprofit",
    // 5: "takeprofitlimit",
    // 6: "stoplosstakeprofit",
    // 7: "stoplosstakeprofitlimit",
    // 8: "trailingstoploss",
    // 9: "trailingstoplosslimit",
    // 10: "stoplossandlimit",
    // 11: "fillorkill",
    // 12: "settleposition"
};
exports.orderTypeToProto = {
    market: 0,
    limit: 1
    // stoploss: 2,
    // stoplosslimit: 3,
    // takeprofit: 4,
    // takeprofitlimit: 5,
    // stoplosstakeprofit: 6,
    // stoplosstakeprofitlimit: 7,
    // trailingstoploss: 8,
    // trailingstoplosslimit: 9,
    // stoplossandlimit: 10,
    // fillorkill: 11,
    // settleposition: 12
};
exports.periodNames = {
    '60': '1m',
    '180': '3m',
    '300': '5m',
    '900': '15m',
    '1800': '30m',
    '3600': '1h',
    '7200': '2h',
    '14400': '4h',
    '21600': '6h',
    '43200': '12h',
    '86400': '1d',
    '259200': '3d',
    '604800': '1w_Thursday',
    '604800_Monday': '1w_Monday'
};
