// cw-sdk-node/build/websocket/proto/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pairUpdateFromProto_1 = require("./pairUpdateFromProto");
const marketUpdateFromProto_1 = require("./marketUpdateFromProto");
const placeOrderOptToProto_1 = require("./placeOrderOptToProto");
const privateOrderFromProto_1 = require("./privateOrderFromProto");
const privatePositionFromProto_1 = require("./privatePositionFromProto");
const privateTradeFromProto_1 = require("./privateTradeFromProto");
const keyToStreamSubscription_1 = require("./keyToStreamSubscription");
const tradeSubscriptionToProto_1 = require("./tradeSubscriptionToProto");
const subscriptionResultFromProto_1 = require("./subscriptionResultFromProto");
const proto = {
    pairUpdateFromProto: pairUpdateFromProto_1.pairUpdateFromProto,
    marketUpdateFromProto: marketUpdateFromProto_1.marketUpdateFromProto,
    placeOrderOptToProto: placeOrderOptToProto_1.placeOrderOptToProto,
    privateOrderFromProto: privateOrderFromProto_1.privateOrderFromProto,
    privateTradeFromProto: privateTradeFromProto_1.privateTradeFromProto,
    privatePositionFromProto: privatePositionFromProto_1.privatePositionFromProto,
    keyToStreamSubscription: keyToStreamSubscription_1.keyToStreamSubscription,
    tradeSubscriptionToProto: tradeSubscriptionToProto_1.tradeSubscriptionToProto,
    subscriptionResultFromProto: subscriptionResultFromProto_1.subscriptionResultFromProto
};
exports.default = proto;
