// cw-sdk-node/build/websocket/proto/pairUpdateFromProto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pairUpdateFromProto = void 0;
const helpers_1 = require("../../util/helpers");
function pairUpdateFromProto(pairUpdate) {
    if (pairUpdate == null || pairUpdate.pair == null) {
        return null;
    }
    const pair = {
        id: helpers_1.getNumber(pairUpdate.pair)
    };
    if (pairUpdate.vwapUpdate && pairUpdate.vwapUpdate.vwap && pairUpdate.vwapUpdate.timestamp) {
        const vwapUpdate = {
            vwap: String(pairUpdate.vwapUpdate.vwap),
            timestamp: helpers_1.getDateFromSecs(pairUpdate.vwapUpdate.timestamp)
        };
        return {
            pair,
            vwapUpdate
        };
    }
    if (pairUpdate.performanceUpdate &&
        pairUpdate.performanceUpdate.window &&
        pairUpdate.performanceUpdate.performance) {
        const performanceUpdate = {
            window: pairUpdate.performanceUpdate.window,
            performance: String(pairUpdate.performanceUpdate.performance)
        };
        return {
            pair,
            performanceUpdate
        };
    }
    if (pairUpdate.trendlineUpdate &&
        pairUpdate.trendlineUpdate.window &&
        pairUpdate.trendlineUpdate.time &&
        pairUpdate.trendlineUpdate.price &&
        pairUpdate.trendlineUpdate.volume) {
        const trendlineUpdate = {
            window: pairUpdate.trendlineUpdate.window,
            timestamp: helpers_1.getDateFromSecs(pairUpdate.trendlineUpdate.time),
            price: pairUpdate.trendlineUpdate.price,
            volume: pairUpdate.trendlineUpdate.volume
        };
        return {
            pair,
            trendlineUpdate
        };
    }
    return null;
}
exports.pairUpdateFromProto = pairUpdateFromProto;
