// cw-sdk-node/build/websocket/proto/privatePositionFromProto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.privatePositionFromProto = void 0;
const helpers_1 = require("../../util/helpers");
const sideFromProto_1 = require("./sideFromProto");
function privatePositionFromProto(position) {
    if (position.id == null ||
        position.time == null ||
        position.side == null ||
        position.avgPriceString == null ||
        position.amountOpenString == null ||
        position.amountClosedString == null ||
        !Array.isArray(position.orderIds) ||
        !Array.isArray(position.tradeIds)) {
        return null;
    }
    const side = sideFromProto_1.privatePositionSideFromProto(position.side);
    if (!side) {
        return null;
    }
    return {
        id: position.id,
        timestamp: helpers_1.getDateFromSecs(position.time),
        side,
        avgPrice: position.avgPriceString,
        amountOpen: position.amountOpenString,
        amountClosed: position.amountClosedString,
        orderIDs: position.orderIds,
        tradeIDs: position.tradeIds
    };
}
exports.privatePositionFromProto = privatePositionFromProto;
