// cw-sdk-node/build/websocket/proto/privateTradeFromProto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.privateTradeFromProto = void 0;
const helpers_1 = require("../../util/helpers");
const sideFromProto_1 = require("./sideFromProto");
function privateTradeFromProto(trade) {
    if (trade.id == null ||
        trade.externalId == null ||
        trade.orderId == null ||
        trade.side == null ||
        !trade.priceString ||
        !trade.amountString) {
        return null;
    }
    let d;
    if (trade.timeMillis && trade.timeMillis > 0) {
        d = helpers_1.getDateFromMs(trade.timeMillis);
    }
    else if (trade.time != null) {
        d = helpers_1.getDateFromSecs(trade.time);
    }
    else {
        return null;
    }
    const side = sideFromProto_1.privateTradeSideFromProto(trade.side);
    if (!side) {
        return null;
    }
    return {
        id: trade.externalId,
        orderID: trade.orderId,
        timestamp: d,
        side,
        price: trade.priceString,
        amount: trade.amountString
    };
}
exports.privateTradeFromProto = privateTradeFromProto;
