// cw-sdk-node/build/util/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/* eslint-disable @typescript-eslint/no-explicit-any */
const winston_1 = tslib_1.__importDefault(require("winston"));
const levels = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
    disabled: 4
};
const defaultLevel = 'info';
class Logger {
    // Constructor is only used in getInstance
    constructor(level) {
        if (!level) {
            level = defaultLevel;
        }
        this.level = level;
        const loggerFormat = winston_1.default.format.printf((info) => {
            return `[${info.label}] ${info.message}`;
        });
        this.logger = winston_1.default.createLogger({
            format: winston_1.default.format.combine(winston_1.default.format.label({ label: 'cw-sdk-node' }), winston_1.default.format.colorize({ all: true }), winston_1.default.format.splat(), winston_1.default.format.simple(), loggerFormat),
            transports: [
                new winston_1.default.transports.Console({
                    level: 'debug'
                })
            ]
        });
    }
    setLevel(newLevel) {
        this.level = newLevel || defaultLevel;
    }
    disable() {
        this.level = 'disabled';
    }
    error(message, ...meta) {
        if (this.shouldLog('error')) {
            this.logger.error(message, meta);
        }
    }
    warn(message, ...meta) {
        if (this.shouldLog('warn')) {
            this.logger.warn(message, meta);
        }
    }
    info(message, ...meta) {
        if (this.shouldLog('info')) {
            this.logger.info(message, meta);
        }
    }
    debug(message, ...meta) {
        if (this.shouldLog('debug')) {
            this.logger.debug(message, meta);
        }
    }
    shouldLog(level) {
        return levels[this.level] <= levels[level];
    }
}
const logger = new Logger(defaultLevel);
exports.default = logger;
