// cw-sdk-node/build/websocket/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventBalancesUpdate = exports.EventPositionsUpdate = exports.EventTradesUpdate = exports.EventOrdersUpdate = exports.DataPerformanceUpdate = exports.DataVWAPUpdate = exports.EventSubscriptionResult = exports.EventUnsubscriptionResult = exports.EventPairUpdate = exports.EventMarketUpdate = exports.EventWSAuthResult = exports.EventWSData = exports.EventStateChange = exports.EventClientError = exports.StateWaitingToReconnect = exports.StateConnected = exports.StateConnecting = exports.StateDisconnected = void 0;
// Connection state change events
exports.StateDisconnected = Symbol('disconnected');
exports.StateConnecting = Symbol('connecting');
exports.StateConnected = Symbol('connected');
exports.StateWaitingToReconnect = Symbol('disconnected: waiting to reconnect');
// WebSocket events
exports.EventClientError = Symbol('client error');
exports.EventStateChange = Symbol('state change');
exports.EventWSData = Symbol('websocket data');
exports.EventWSAuthResult = Symbol('auth result');
// Stream Client events
exports.EventMarketUpdate = Symbol('market update');
exports.EventPairUpdate = Symbol('pair update');
exports.EventUnsubscriptionResult = Symbol('unsubscription result');
exports.EventSubscriptionResult = Symbol('subscription result');
exports.DataVWAPUpdate = Symbol('vwap update');
exports.DataPerformanceUpdate = Symbol('performance update');
// Trade Client
exports.EventOrdersUpdate = Symbol('orders update');
exports.EventTradesUpdate = Symbol('trades update');
exports.EventPositionsUpdate = Symbol('positions update');
exports.EventBalancesUpdate = Symbol('balances update');
